pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -836,37 +839,36 @@ func (c *Controller) collectAllForeignEndpoints(svc *model.Service) []*model.Ist
 		return nil
 	}
 
-	instances := c.getForeignServiceInstancesByPort(svc, svc.Ports[0].Port)
 	endpoints := make([]*model.IstioEndpoint, 0)
-
-	// all endpoints for ports[0]
-	for _, instance := range instances {
-		endpoints = append(endpoints, instance.Endpoint)
-	}
-
-	// build an endpoint for each remaining service port
-	for i := 1; i < len(svc.Ports); i++ {
+	for _, port := range svc.Ports {
+		instances := c.getForeignServiceInstancesByPort(svc, port.Port)
 		for _, instance := range instances {
-			ep := *instance.Endpoint
-			ep.EndpointPort = uint32(svc.Ports[i].Port)
-			ep.ServicePortName = svc.Ports[i].Name
-			endpoints = append(endpoints, &ep)
```

Comment:
I explicitly avoided calling the getForeignServiceInstancesByPort to reduce the lock contention. We are taking the global controller lock everytime we call the above function. And the only thing changing across each call is the endpoint's port and port name.